lib/std/math.zig

```diff
@@ -36,33 +36,39 @@ pub const sqrt2 = 1.414213562373095048801688724209698079;
 /// 1/sqrt(2)
 pub const sqrt1_2 = 0.707106781186547524400844362104849039;
 
-// From a small c++ [program using boost float128](https://github.com/winksaville/cpp_boost_float128)
-pub const f128_true_min = @bitCast(f128, @as(u128, 0x00000000000000000000000000000001));
-pub const f128_min = @bitCast(f128, @as(u128, 0x00010000000000000000000000000000));
-pub const f128_max = @bitCast(f128, @as(u128, 0x7FFEFFFFFFFFFFFFFFFFFFFFFFFFFFFF));
-pub const f128_epsilon = @bitCast(f128, @as(u128, 0x3F8F0000000000000000000000000000));
-pub const f128_toint = 1.0 / f128_epsilon;
-
-// float.h details
-pub const f64_true_min = 4.94065645841246544177e-324;
-pub const f64_min = 2.2250738585072014e-308;
-pub const f64_max = 1.79769313486231570815e+308;
-pub const f64_epsilon = 2.22044604925031308085e-16;
-pub const f64_toint = 1.0 / f64_epsilon;
-
-pub const f32_true_min = 1.40129846432481707092e-45;
-pub const f32_min = 1.17549435082228750797e-38;
-pub const f32_max = 3.40282346638528859812e+38;
-pub const f32_epsilon = 1.1920928955078125e-07;
-pub const f32_toint = 1.0 / f32_epsilon;
-
-pub const f16_true_min = 0.000000059604644775390625; // 2**-24
-pub const f16_min = 0.00006103515625; // 2**-14
-pub const f16_max = 65504;
-pub const f16_epsilon = 0.0009765625; // 2**-10
+pub const floatExponentBits = @import("math/float.zig").floatExponentBits;
+pub const floatMantissaBits = @import("math/float.zig").floatMantissaBits;
+pub const floatExponentMin = @import("math/float.zig").floatExponentMin;
+pub const floatExponentMax = @import("math/float.zig").floatExponentMax;
+pub const floatTrueMin = @import("math/float.zig").floatTrueMin;
+pub const floatMin = @import("math/float.zig").floatMin;
+pub const floatMax = @import("math/float.zig").floatMax;
+pub const floatEps = @import("math/float.zig").floatEps;
+
+// These should eventually be removed, see #10133
+pub const f16_true_min = floatTrueMin(f16);
+pub const f32_true_min = floatTrueMin(f32);
+pub const f64_true_min = floatTrueMin(f64);
+pub const f128_true_min = floatTrueMin(f128);
+pub const f16_min = floatMin(f16);
+pub const f32_min = floatMin(f32);
+pub const f64_min = floatMin(f64);
+pub const f128_min = floatMin(f128);
+pub const f16_max = floatMax(f16);
+pub const f32_max = floatMax(f32);
+pub const f64_max = floatMax(f64);
+pub const f128_max = floatMax(f128);
+pub const f16_epsilon = floatEps(f16);
+pub const f32_epsilon = floatEps(f32);
+pub const f64_epsilon = floatEps(f64);
+pub const f128_epsilon = floatEps(f128);
 pub const f16_toint = 1.0 / f16_epsilon;
-
-pub const epsilon = @import("math/epsilon.zig").epsilon;
+pub const f32_toint = 1.0 / f32_epsilon;
+pub const f64_toint = 1.0 / f64_epsilon;
+pub const f128_toint = 1.0 / f128_epsilon;
+pub fn epsilon(comptime T: type) T {
+    return floatEps(T);
+}
```

Comment:
Is there a reason for making the float functions return `comptime_float` rather than `T`? I would expect the constants above to be the type that's in the name (e.g. `f32_max` of type `f32`), and then `epsilon()` could just be an alias for `floatEps()` (planned to be removed?).

Reply:
I went with what std currently does:

- `epsilon` coerces to `T`, not matching this behaviour caused some random testcases to flip out
- the other ones are `comptime_float`

I also checked if all of them coerce correctly to their concrete types, which they appear to do. Do you think I should make them all `T`, and have the compatibility constants re-coerce to `comptime_float`?

Reply:
Ah I hadn't thought that the constants were already `comptime_float`, that should've been obvious to me! Do things break if the constants are made `T` instead of `comptime_float`? (or is there some other subtle reason for making them comptime that I'm overlooking?). I'm happy with what you have now, was just wondering really :)

Reply:
> Do things break if they're made `T`?

Surprisingly yes, see [this CI build](https://ci.ziglang.org/ziglang/zig/1009/1/4).

I just noticed however that the `f128` constants are in fact *not* comptime due to being consteucted from bit repr, so maybe it'd be best to go with `T` and do `@as(comptime_float, ...)` for the other constants, as I do think it makes more sense to receive a `T` too, I think pretty much every usecase would be coercing them to `T` for usage.

Reply:
The CI failure you linked seems to be a problem with `epsilon()` returning a `comptime_float`, not with the constants being type `T` (e.g. `f32`)? FWIW the problem is that multiplying a `comptime_float` and a `comptime_int` gives a compile error - relevant issue: #8334. Have you seen failures when making the consts `T`, as opposed to this error with `epsilon()` being `comptime_float`?

Reply:
You're right, that's kind of the opposite issue. I don't believe I saw anything break with the mins and maxes since it wouldn't be hard to imagine it breaking a simple `@TypeOf(x)` for someone so I just kept this PR entirely non-breaking.

Reply:
Why would anyone be doing a `@TypeOf(f64_min)` and expecting `comptime_float`, given `f64` is even in the name? Thinking about it I do feel like the constants should be made `T` rather than `comptime_float` - as you point out the `f128` constants already are, so this would give consistency with that. I suspect this was more something that was overlooked rather than intentional.

Reply:
> I suspect this was more something that was overlooked rather than intentional.

I have the same feeling from `f128` being a concrete type while the rest just aren't. I'm probably overthinking it, I'll change the constants.

Reply:
Nevermind, it breaks [compiler_rt](https://ci.ziglang.org/ziglang/zig/1014/1/4) of all things, definitely not touching that. I'll make the old constants be what they were before.